File: experiments/009-CuisByScripts/cells/java-gradle_1-Empty_free/2024-1c-parcial1/20260916-095748-7943/workdir/src/main/java/simshotels/Hotel.java
package simshotels;

import java.util.ArrayList;
import java.util.List;
import java.util.function.ToIntFunction;

public class Hotel {

    private final List<Floor> floors;

    // instance creation (class side)

    public static Hotel withFloors(List<Floor> aFloorsCollection) {
        Hotel.assertHaveFloors(aFloorsCollection);

        return new Hotel(aFloorsCollection);
    }

    private Hotel(List<Floor> aFloorsCollection) {
        floors = new ArrayList<>(aFloorsCollection);
    }

    // accounting

    public int totalLosses() {
        return sumOverFloors(Floor::totalLosses);
    }

    public int totalProfits() {
        return sumOverFloors(Floor::totalProfits);
    }

    // totals

    public int totalRooms() {
        return sumOverFloors(Floor::totalRooms);
    }

    public int totalRoomsAvailable() {
        return sumOverFloors(Floor::totalRoomsAvailable);
    }

    public int totalRoomsOccupied() {
        return sumOverFloors(Floor::totalRoomsOccupied);
    }

    public int totalRoomsReserved() {
        return sumOverFloors(Floor::totalRoomsReserved);
    }

    private int sumOverFloors(ToIntFunction<Floor> aFloorValue) {
        return floors.stream().mapToInt(aFloorValue).sum();
    }

    // testing

    public boolean isEmpty() {
        return floors.stream().allMatch(Floor::isAvailable);
    }

    // guests

    public void receiveAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        floorNumbered(aFloorNumber).receiveAtRoom(aGuestType, aRoomNumber);
    }

    public void receiveWithReservationAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        floorNumbered(aFloorNumber).receiveWithReservationAtRoom(aGuestType, aRoomNumber);
    }

    public void reserveRoomAtFloor(int aRoomNumber, int aFloorNumber) {
        floorNumbered(aFloorNumber).reserveRoom(aRoomNumber);
    }

    private Floor floorNumbered(int aFloorNumber) {
        if (aFloorNumber < 1 || aFloorNumber > floors.size()) {
            Hotel.signalFloorNumberDoesNotExist();
        }
        return floors.get(aFloorNumber - 1);
    }

    // assertions (class side)

    public static void assertHaveFloors(List<Floor> aFloorsCollection) {
        if (aFloorsCollection.isEmpty()) {
            Hotel.signalNoFloors();
        }
    }

    // error description (class side)

    public static String floorNumberDoesNotExistErrorDescription() {
        return "Floor number does not exist";
    }

    public static String noFloorsErrorDescription() {
        return "Cannot have a Hotel without floors";
    }

    // exceptions (class side)

    public static void signalFloorNumberDoesNotExist() {
        throw new RuntimeException(Hotel.floorNumberDoesNotExistErrorDescription());
    }

    public static void signalNoFloors() {
        throw new RuntimeException(Hotel.noFloorsErrorDescription());
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2024-1c-parcial1/20260921-203459-63160/output/project/src/simshotels/hotel.py
from simshotels.floor import Floor


class Hotel:

    # instance creation (class side)

    @classmethod
    def with_floors(cls, a_floors_collection):
        Hotel.assert_have_floors(a_floors_collection)

        return cls(a_floors_collection)

    # initialization

    def __init__(self, a_floors_collection):
        self._floors = a_floors_collection

    # accounting

    def total_losses(self):
        return sum(floor.total_losses() for floor in self._floors)

    def total_profits(self):
        return sum(floor.total_profits() for floor in self._floors)

    # totals

    def total_rooms(self):
        return sum(floor.total_rooms() for floor in self._floors)

    def total_rooms_available(self):
        return sum(floor.total_rooms_available() for floor in self._floors)

    def total_rooms_occupied(self):
        return sum(floor.total_rooms_occupied() for floor in self._floors)

    def total_rooms_reserved(self):
        return sum(floor.total_rooms_reserved() for floor in self._floors)

    # testing

    def is_empty(self):
        return all(floor.is_available() for floor in self._floors)

    # guests

    def receive_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        self.floor_at(a_floor_number).receive_at_room(a_guest_type, a_room_number)

    def receive_with_reservation_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        self.floor_at(a_floor_number).receive_with_reservation_at_room(a_guest_type, a_room_number)

    def reserve_room_at_floor(self, a_room_number, a_floor_number):
        self.floor_at(a_floor_number).reserve_room(a_room_number)

    # private

    def floor_at(self, a_floor_number):
        if a_floor_number < 1 or a_floor_number > len(self._floors):
            Hotel.signal_floor_number_does_not_exist()

        return self._floors[a_floor_number - 1]

    # assertions (class side)

    @staticmethod
    def assert_have_floors(a_floors_collection):
        if len(a_floors_collection) == 0:
            Hotel.signal_no_floors()

    # error description (class side)

    @staticmethod
    def floor_number_does_not_exist_error_description():
        return "Floor number does not exist"

    @staticmethod
    def no_floors_error_description():
        return "Cannot have a Hotel without floors"

    @staticmethod
    def room_number_does_not_exist_error_description():
        # room numbers are known by the floor, the hotel just publishes its error description
        return Floor.room_number_does_not_exist_error_description()

    # exceptions (class side)

    @staticmethod
    def signal_floor_number_does_not_exist():
        raise RuntimeError(Hotel.floor_number_does_not_exist_error_description())

    @staticmethod
    def signal_no_floors():
        raise RuntimeError(Hotel.no_floors_error_description())
